tokio/src/task/local.rs

```diff
@@ -0,0 +1,585 @@
+/// use std::rc::Rc;
+/// use tokio::task;
+///
+/// let unsend_data = Rc::new("my unsend data...");
+///
+/// let mut rt = Runtime::new().unwrap();
+/// // Construct a local task set that can run `!Send` futures.
+/// let local = task::LocalSet::new();
+///
+/// // Run the local task group.
+/// local.block_on(&mut rt, async move {
+///     let unsend_data = unsend_data.clone();
+///     // `spawn_local` ensures that the future is spawned on the local
+///     // task group.
+///     task::spawn_local(async move {
+///         println!("{}", unsend_data);
+///         // ...
+///     }).await.unwrap();
+/// });
+/// ```
+///
+/// [`Send`]: https://doc.rust-lang.org/std/marker/trait.Send.html
+/// [local task set]: struct.LocalSet.html
+/// [`Runtime::block_on`]: ../struct.Runtime.html#method.block_on
+/// [`task::spawn_local`]: fn.spawn.html
+#[derive(Debug)]
+pub struct LocalSet {
+    scheduler: Rc<Scheduler>,
+    _not_send_or_sync: PhantomData<*const ()>,
+}
+
+struct Scheduler {
+    /// List of all active tasks spawned onto this executor.
+    ///
+    /// # Safety
+    ///
+    /// Must only be accessed from the primary thread
+    tasks: UnsafeCell<task::OwnedList<Scheduler, Unsendable>>,
+
+    /// Local run queue.
+    ///
+    /// Tasks notified from the current thread are pushed into this queue.
+    ///
+    /// # Safety
+    ///
+    /// References should not be handed out. Only call `push` / `pop` functions.
+    /// Only call from the owning thread.
+    queue: UnsafeCell<VecDeque<UnsendTask<Scheduler>>>,
+}
+
+#[pin_project]
+struct LocalFuture<F> {
+    scheduler: Rc<Scheduler>,
+    #[pin]
+    future: F,
+}
+
+thread_local! {
+    static CURRENT_TASK_SET: Cell<Option<NonNull<Scheduler>>> = Cell::new(None);
+}
+
+/// Spawns a `!Send` future on the local task set.
+///
+/// The spawned future will be run on the same thread that called `spawn_local.`
+/// This may only be called from the context of a local task set.
+///
+/// # Panics
+///
+/// - This function panics if called outside of a local task set.
+///
+/// # Examples
+///
+/// ```rust
+/// # use tokio::runtime::Runtime;
+/// use std::rc::Rc;
+/// use tokio::task;
+///
+/// let unsend_data = Rc::new("my unsend data...");
+///
+/// let mut rt = Runtime::new().unwrap();
+/// let local = task::LocalSet::new();
+///
+/// // Run the local task set.
+/// local.block_on(&mut rt, async move {
+///     let unsend_data = unsend_data.clone();
+///     task::spawn_local(async move {
+///         println!("{}", unsend_data);
+///         // ...
+///     }).await.unwrap();
+/// });
+/// ```
+pub fn spawn_local<F>(future: F) -> JoinHandle<F::Output>
+where
+    F: Future + 'static,
+    F::Output: 'static,
+{
+    CURRENT_TASK_SET.with(|current| {
+        let current = current
+            .get()
+            .expect("`spawn_local` called from outside of a local::LocalSet!");
+        unsafe {
+            let (task, handle) = task::joinable_unsend(future);
```

Comment:
we should probably stick w/ `local` instead of `unsend` since we are doubling down on that.